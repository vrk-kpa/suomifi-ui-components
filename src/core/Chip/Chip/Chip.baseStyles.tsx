import { css } from 'styled-components';
import { SuomifiTheme } from '../../theme';
import { baseChipBaseStyles } from '../BaseChip/BaseChip.baseStyles';
import { MarginProps, buildSpacingCSS } from '../../theme/utils/spacing';

export const baseStyles = (
  theme: SuomifiTheme,
  globalMargins?: MarginProps,
  propMargins?: MarginProps,
) => css`
  ${baseChipBaseStyles(theme)}
  ${buildSpacingCSS(globalMargins)}
  ${buildSpacingCSS(propMargins, true)}
  &.fi-chip--button {
    cursor: pointer;
    &:hover {
      background: ${theme.colors.highlightLight1};
    }

    &:active {
      background: ${theme.colors.highlightDark1};
    }
    &:focus {
      ${theme.focuses.highContrastFocus}
    }
  }

  &.fi-chip--removable {
    padding-top: ${theme.spacing.insetXxs};
    padding-right: 22px;
    padding-bottom: ${theme.spacing.insetXxs};
    padding-left: ${theme.spacing.insetL};
    position: relative;

    & .fi-chip--icon {
      position: absolute;
      top: 8px;
      right: 10px;
      height: 12px;
      width: 12px;
    }

    & .fi-chip--content {
      margin-right: ${theme.spacing.xs};
    }

    &.fi-chip--disabled {
      & .fi-chip--icon {
        cursor: not-allowed;
      }
    }
  }
`;
