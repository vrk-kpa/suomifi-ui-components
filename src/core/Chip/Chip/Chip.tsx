import React, { Component, forwardRef } from 'react';
import classnames from 'classnames';
import { default as styled } from 'styled-components';
import { IconClose } from 'suomifi-icons';
import { getLogger } from '../../../utils/log';
import { HtmlButton, HtmlButtonProps, HtmlSpan } from '../../../reset';
import {
  separateMarginProps,
  MarginProps,
  GlobalMarginProps,
} from '../../theme/utils/spacing';
import { VisuallyHidden } from '../../VisuallyHidden/VisuallyHidden';
import {
  BaseChipProps,
  baseClassName,
  chipClassNames,
} from '../BaseChip/BaseChip';
import { baseStyles } from './Chip.baseStyles';
import {
  SuomifiThemeProp,
  SuomifiThemeConsumer,
  SpacingConsumer,
} from '../../theme';
import { filterDuplicateKeys } from '../../../utils/common/common';

const chipButtonClassNames = {
  removable: `${baseClassName}--removable`,
  button: `${baseClassName}--button`,
  icon: `${baseClassName}--icon`,
};

export interface ChipProps
  extends BaseChipProps,
    MarginProps,
    Omit<
      HtmlButtonProps,
      'forwardedRef' | 'disabled' | 'onClick' | 'children' | 'as'
    > {
  /**
   * Callback fired when Chip is clicked
   */
  onClick?: () => void;
  /**
   * Shows an X-icon next to the content to mark the Chip as removable. Always use `actionLabel` for screen readers when removable is true.
   * @default false
   */
  removable?: boolean;
  /** Creates an aria-label text to let screen reader users know pressing the button will remove the chip/selection.
   * Required with `removable`
   */
  actionLabel?: string;
  /** Soft disables the chip to allow tab-focus. Disables `onClick()` functionality */
  'aria-disabled'?: boolean;
  /** Ref is forwarded to the button element. Alternative to React `ref` attribute. */
  forwardedRef?: React.RefObject<HTMLButtonElement>;
}
class DefaultChip extends Component<ChipProps> {
  render() {
    const {
      className,
      children,
      onClick,
      removable,
      actionLabel,
      forwardedRef,
      disabled = false,
      'aria-disabled': ariaDisabled = false,
      style,
      ...rest
    } = this.props;
    const [_marginProps, passProps] = separateMarginProps(rest);

    const onClickProp = !!disabled || !!ariaDisabled ? {} : { onClick };

    if (removable && !actionLabel) {
      getLogger().error(
        'Provide actionLabel to communicate removability to screen readers',
      );
    }
    return (
      <HtmlButton
        className={classnames(
          baseClassName,
          chipButtonClassNames.button,
          className,
          {
            [chipClassNames.disabled]: !!disabled || !!ariaDisabled,
            [chipButtonClassNames.removable]: !!removable,
          },
        )}
        disabled={disabled}
        aria-disabled={!!ariaDisabled || !!disabled}
        {...onClickProp}
        forwardedRef={forwardedRef}
        {...passProps}
        style={style}
      >
        <HtmlSpan className={chipClassNames.content}>{children}</HtmlSpan>
        {!!removable && (
          <>
            <IconClose
              mousePointer={true}
              color="currentColor"
              className={chipButtonClassNames.icon}
              aria-hidden={true}
            />
            <VisuallyHidden>{actionLabel}</VisuallyHidden>
          </>
        )}
      </HtmlButton>
    );
  }
}

const StyledChip = styled(
  ({
    theme,
    globalMargins,
    ...passProps
  }: ChipProps & SuomifiThemeProp & GlobalMarginProps) => (
    <DefaultChip {...passProps} />
  ),
)`
  ${({ theme, globalMargins, ...rest }) => {
    const [marginProps, _passProps] = separateMarginProps(rest);
    const cleanedGlobalMargins = filterDuplicateKeys(
      globalMargins.chip,
      marginProps,
    );
    return baseStyles(theme, cleanedGlobalMargins, marginProps);
  }}
`;

const Chip = forwardRef(
  (props: ChipProps, ref: React.RefObject<HTMLButtonElement>) => {
    const { ...passProps } = props;
    return (
      <SpacingConsumer>
        {({ margins }) => (
          <SuomifiThemeConsumer>
            {({ suomifiTheme }) => (
              <StyledChip
                theme={suomifiTheme}
                globalMargins={margins}
                forwardedRef={ref}
                {...passProps}
              />
            )}
          </SuomifiThemeConsumer>
        )}
      </SpacingConsumer>
    );
  },
);

Chip.displayName = 'Chip';
export { Chip };
