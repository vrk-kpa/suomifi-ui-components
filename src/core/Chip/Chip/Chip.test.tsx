import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { Chip } from './Chip';

describe('disabled', () => {
  const DisabledChip = (
    <Chip disabled={true} onClick={() => null} actionLabel="Unselect">
      Testcontent
    </Chip>
  );

  it('has "--disabled"-class', () => {
    const { container } = render(DisabledChip);
    expect(container.firstChild).toHaveClass('fi-chip--disabled');
  });

  it('will add "disabled"-attribute for the button element', () => {
    const { getByRole } = render(DisabledChip);
    expect(getByRole('button')).toHaveAttribute('disabled');
  });

  it('is aria-disabled and changes to enabled', () => {
    const { getByRole, rerender } = render(
      <Chip aria-disabled={true}>Aria-disabled Chip</Chip>,
    );
    expect(getByRole('button')).toHaveAttribute('aria-disabled', 'true');
    rerender(<Chip aria-disabled={false}>Aria-disabled Chip</Chip>);
    expect(getByRole('button')).toHaveAttribute('aria-disabled', 'false');
  });

  it('should match snapshot', () => {
    const { container } = render(DisabledChip);
    expect(container).toMatchSnapshot();
  });
});

describe('children', () => {
  const chipWithChildren = (
    <Chip>
      <div data-testid="test-div">Selection 1</div>
    </Chip>
  );

  it('has the given content', () => {
    const { getByTestId } = render(chipWithChildren);
    expect(getByTestId('test-div').textContent).toBe('Selection 1');
  });

  it('should match snapshot', () => {
    const { container } = render(chipWithChildren);
    expect(container).toMatchSnapshot();
  });
});

describe('classnames', () => {
  const customClassChip = <Chip className="custom-class">Testcontent</Chip>;

  it('contains base classname', () => {
    const { container } = render(customClassChip);
    expect(container.firstChild).toHaveClass('fi-chip');
  });

  it('contains custom classname', () => {
    const { container } = render(customClassChip);
    expect(container.firstChild).toHaveClass('custom-class');
  });

  it('should match snapshot', () => {
    const { container } = render(customClassChip);
    expect(container).toMatchSnapshot();
  });
});

describe('variant specifics', () => {
  const defaultChip = (
    <Chip removable={true} onClick={() => null} actionLabel="Unselect">
      Test content
    </Chip>
  );

  it('has icon element when set as removable', () => {
    const { container } = render(defaultChip);
    const icon = container.querySelector('.fi-chip--icon');
    expect(container.contains(icon)).toBeTruthy();
  });
});

describe('onClick', () => {
  it('is called when clicked', () => {
    const mockClick = jest.fn();
    const { getByRole } = render(
      <Chip onClick={mockClick} actionLabel="Unselect">
        Test content
      </Chip>,
    );
    const chip = getByRole('button');
    fireEvent.click(chip);
    expect(mockClick).toHaveBeenCalledTimes(1);
  });

  describe('margin prop', () => {
    it('should have margin style from margin prop', () => {
      const { container } = render(<Chip margin="xs">Test</Chip>);
      expect(container.firstChild).toHaveStyle('margin: 10px');
    });

    it('should have margin prop style overwritten from style', () => {
      const { container } = render(
        <Chip margin="xs" style={{ margin: 2 }}>
          Test
        </Chip>,
      );
      expect(container.firstChild).toHaveAttribute('style', 'margin: 2px;');
    });
  });
});
