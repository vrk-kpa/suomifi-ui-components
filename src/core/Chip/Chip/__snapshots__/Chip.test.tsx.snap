// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`children should match snapshot 1`] = `
.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  overflow: visible;
  text-transform: none;
  -webkit-appearance: button;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline-block;
  max-width: 100%;
  cursor: pointer;
}

.c0:-moz-focusring {
  outline: 1px dotted ButtonText;
}

.c0::-moz-focus-inner {
  border-style: none;
  padding: 0;
}

.c0::-webkit-file-upload-button {
  -webkit-appearance: button;
  font: inherit;
}

.c0::-webkit-inner-spin-button {
  height: auto;
}

.c0::-webkit-outer-spin-button {
  height: auto;
}

.c0::before,
.c0::after {
  box-sizing: border-box;
}

.c2 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c2::before,
.c2::after {
  box-sizing: border-box;
}

.c1 {
  color: hsl(0,0%,13%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
}

.c1.fi-chip {
  border-radius: 14px;
  padding: 2px 10px;
  color: hsl(0,0%,100%);
  background: hsl(212,63%,45%);
  max-height: 100%;
  display: inline-block;
}

.c1.fi-chip:focus {
  outline: 0;
  position: relative;
}

.c1.fi-chip:focus::after {
  content: '';
  position: absolute;
  pointer-events: none;
  top: -2px;
  right: -2px;
  bottom: -2px;
  left: -2px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0,0%,100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
  z-index: 9999;
  border-radius: 16px;
}

.c1.fi-chip .fi-chip--content {
  display: block;
  word-break: break-word;
  overflow: hidden;
  line-height: 1.5em;
}

.c1.fi-chip--disabled.fi-chip {
  cursor: not-allowed;
  background: hsl(202,7%,67%);
}

.c1.fi-chip--disabled.fi-chip:hover,
.c1.fi-chip--disabled.fi-chip:active {
  background: hsl(202,7%,67%);
}

.c1.fi-chip--button {
  cursor: pointer;
}

.c1.fi-chip--button:hover {
  background: hsl(212,63%,49%);
}

.c1.fi-chip--button:active {
  background: hsl(212,63%,37%);
}

.c1.fi-chip--button:focus {
  outline: 3px solid transparent;
}

.c1.fi-chip--removable {
  padding-top: 2px;
  padding-right: 22px;
  padding-bottom: 2px;
  padding-left: 10px;
  position: relative;
}

.c1.fi-chip--removable .fi-chip--icon {
  position: absolute;
  top: 8px;
  right: 10px;
  height: 12px;
  width: 12px;
}

.c1.fi-chip--removable .fi-chip--content {
  margin-right: 10px;
}

.c1.fi-chip--removable.fi-chip--disabled .fi-chip--icon {
  cursor: not-allowed;
}

@media screen and (forced-colors:active) {
  .c1.fi-chip--disabled.fi-chip {
    color: GrayText;
  }
}

<div>
  <button
    aria-disabled="false"
    class="c0 fi-chip fi-chip--button c1"
    type="button"
  >
    <span
      class="c2 fi-chip--content"
    >
      <div
        data-testid="test-div"
      >
        Selection 1
      </div>
    </span>
  </button>
</div>
`;

exports[`classnames should match snapshot 1`] = `
.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  overflow: visible;
  text-transform: none;
  -webkit-appearance: button;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline-block;
  max-width: 100%;
  cursor: pointer;
}

.c0:-moz-focusring {
  outline: 1px dotted ButtonText;
}

.c0::-moz-focus-inner {
  border-style: none;
  padding: 0;
}

.c0::-webkit-file-upload-button {
  -webkit-appearance: button;
  font: inherit;
}

.c0::-webkit-inner-spin-button {
  height: auto;
}

.c0::-webkit-outer-spin-button {
  height: auto;
}

.c0::before,
.c0::after {
  box-sizing: border-box;
}

.c2 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c2::before,
.c2::after {
  box-sizing: border-box;
}

.c1 {
  color: hsl(0,0%,13%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
}

.c1.fi-chip {
  border-radius: 14px;
  padding: 2px 10px;
  color: hsl(0,0%,100%);
  background: hsl(212,63%,45%);
  max-height: 100%;
  display: inline-block;
}

.c1.fi-chip:focus {
  outline: 0;
  position: relative;
}

.c1.fi-chip:focus::after {
  content: '';
  position: absolute;
  pointer-events: none;
  top: -2px;
  right: -2px;
  bottom: -2px;
  left: -2px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0,0%,100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
  z-index: 9999;
  border-radius: 16px;
}

.c1.fi-chip .fi-chip--content {
  display: block;
  word-break: break-word;
  overflow: hidden;
  line-height: 1.5em;
}

.c1.fi-chip--disabled.fi-chip {
  cursor: not-allowed;
  background: hsl(202,7%,67%);
}

.c1.fi-chip--disabled.fi-chip:hover,
.c1.fi-chip--disabled.fi-chip:active {
  background: hsl(202,7%,67%);
}

.c1.fi-chip--button {
  cursor: pointer;
}

.c1.fi-chip--button:hover {
  background: hsl(212,63%,49%);
}

.c1.fi-chip--button:active {
  background: hsl(212,63%,37%);
}

.c1.fi-chip--button:focus {
  outline: 3px solid transparent;
}

.c1.fi-chip--removable {
  padding-top: 2px;
  padding-right: 22px;
  padding-bottom: 2px;
  padding-left: 10px;
  position: relative;
}

.c1.fi-chip--removable .fi-chip--icon {
  position: absolute;
  top: 8px;
  right: 10px;
  height: 12px;
  width: 12px;
}

.c1.fi-chip--removable .fi-chip--content {
  margin-right: 10px;
}

.c1.fi-chip--removable.fi-chip--disabled .fi-chip--icon {
  cursor: not-allowed;
}

@media screen and (forced-colors:active) {
  .c1.fi-chip--disabled.fi-chip {
    color: GrayText;
  }
}

<div>
  <button
    aria-disabled="false"
    class="c0 fi-chip fi-chip--button c1 custom-class"
    type="button"
  >
    <span
      class="c2 fi-chip--content"
    >
      Testcontent
    </span>
  </button>
</div>
`;

exports[`disabled should match snapshot 1`] = `
.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  overflow: visible;
  text-transform: none;
  -webkit-appearance: button;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline-block;
  max-width: 100%;
  cursor: pointer;
}

.c0:-moz-focusring {
  outline: 1px dotted ButtonText;
}

.c0::-moz-focus-inner {
  border-style: none;
  padding: 0;
}

.c0::-webkit-file-upload-button {
  -webkit-appearance: button;
  font: inherit;
}

.c0::-webkit-inner-spin-button {
  height: auto;
}

.c0::-webkit-outer-spin-button {
  height: auto;
}

.c0::before,
.c0::after {
  box-sizing: border-box;
}

.c2 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c2::before,
.c2::after {
  box-sizing: border-box;
}

.c1 {
  color: hsl(0,0%,13%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
}

.c1.fi-chip {
  border-radius: 14px;
  padding: 2px 10px;
  color: hsl(0,0%,100%);
  background: hsl(212,63%,45%);
  max-height: 100%;
  display: inline-block;
}

.c1.fi-chip:focus {
  outline: 0;
  position: relative;
}

.c1.fi-chip:focus::after {
  content: '';
  position: absolute;
  pointer-events: none;
  top: -2px;
  right: -2px;
  bottom: -2px;
  left: -2px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0,0%,100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
  z-index: 9999;
  border-radius: 16px;
}

.c1.fi-chip .fi-chip--content {
  display: block;
  word-break: break-word;
  overflow: hidden;
  line-height: 1.5em;
}

.c1.fi-chip--disabled.fi-chip {
  cursor: not-allowed;
  background: hsl(202,7%,67%);
}

.c1.fi-chip--disabled.fi-chip:hover,
.c1.fi-chip--disabled.fi-chip:active {
  background: hsl(202,7%,67%);
}

.c1.fi-chip--button {
  cursor: pointer;
}

.c1.fi-chip--button:hover {
  background: hsl(212,63%,49%);
}

.c1.fi-chip--button:active {
  background: hsl(212,63%,37%);
}

.c1.fi-chip--button:focus {
  outline: 3px solid transparent;
}

.c1.fi-chip--removable {
  padding-top: 2px;
  padding-right: 22px;
  padding-bottom: 2px;
  padding-left: 10px;
  position: relative;
}

.c1.fi-chip--removable .fi-chip--icon {
  position: absolute;
  top: 8px;
  right: 10px;
  height: 12px;
  width: 12px;
}

.c1.fi-chip--removable .fi-chip--content {
  margin-right: 10px;
}

.c1.fi-chip--removable.fi-chip--disabled .fi-chip--icon {
  cursor: not-allowed;
}

@media screen and (forced-colors:active) {
  .c1.fi-chip--disabled.fi-chip {
    color: GrayText;
  }
}

<div>
  <button
    aria-disabled="true"
    class="c0 fi-chip fi-chip--button c1 fi-chip--disabled"
    disabled=""
    type="button"
  >
    <span
      class="c2 fi-chip--content"
    >
      Testcontent
    </span>
  </button>
</div>
`;
